### PR TITLE
Fixed #32363 -- Ensured sys.__interactivehook__ is called in shell

### PR DESCRIPTION
**Pull request by inglesp**
ticket-32363

By default, this means that `readline` is properly registered, so that
`.python_history` is used.

`sys.__interactivehook__` may be set by a `$PYTHONSTARTUP` file.

2021-06-16 update: see https://github.com/django/django/pull/13911#issuecomment-862764723 for a summary of the discussion so far.

### CODE REVIEW

**Comment by inglesp** [comment]
See https://github.com/python/cpython/blob/master/Lib/site.py#L441-L487 for the code that is executed by `sys.__interactivehook__` by default.

File: django/core/management/commands/shell.py [hunk 1]
```diff
@@ -42,25 +42,8 @@ def bpython(self, options):
...
-        else:
-            # We don't have to wrap the following import in a 'try', because
-            # we already know 'readline' was imported successfully.
-            import rlcompleter
-            readline.set_completer(rlcompleter.Completer(imported_objects).complete)
-            # Enable tab completion on systems using libedit (e.g. macOS).
-            # These lines are copied from Python's Lib/site.py.
-            readline_doc = getattr(readline, '__doc__', '')
-            if readline_doc is not None and 'libedit' in readline_doc:
-                readline.parse_and_bind("bind ^I rl_complete")
```
Comment (wkschwartz) [inline django/core/management/commands/shell.py:61]: You appear to have dropped support for macOS.

Comment (inglesp) [reply django/core/management/commands/shell.py:61]: How so?

Comment (wkschwartz) [reply django/core/management/commands/shell.py:61]: Unless I misread (maybe I misread!), you removed the four lines of code above this comment that refers to libedit without adding it back somewhere else.

Comment (inglesp) [reply django/core/management/commands/shell.py:61]: This block of code came from default value of `sys.__interactivehook__` (https://github.com/python/cpython/blob/master/Lib/site.py#L441-L487).  Since we now call `sys.__interactivehook__()`, that block can be removed.

But I've only been able to test on linux, so it would be very helpful if you could test on macOS.

Comment (wkschwartz) [reply django/core/management/commands/shell.py:61]: I see. Anyway this is still a breaking change. The function to which you link, `site.enablerlcompleter`, is called [only if](https://docs.python.org/3/library/site.html#readline-configuration) Python was started [without](https://github.com/python/cpython/blob/eeb701adc0fc29ba803fddf133d917ff45639a00/Lib/site.py#L614-L615) [`-S`] and [without](https://github.com/python/cpython/blob/eeb701adc0fc29ba803fddf133d917ff45639a00/Lib/site.py#L606-L607) [`-I`] (see also [`sys.flags`]), so it doesn't always set `sys.__interactivehook__` by default. The prior behavior of the `shell` command` enabled the rlcompleter with macOS support unconditionally. Your patch now only enables macOS support under some operating conditions.

Perhaps it would be better to lean directly on the `site` module. I'm not sure it's a good idea to call `site.enablerlcompleter`: On the one hand, it's not documented. On the other, the name doesn't start with an underscore. One option might be unconditionally calling [`site.main`](https://docs.python.org/3/library/site.html#site.main) so that user and site customization script behavior matches vanilla Python.

[`-S`]: https://docs.python.org/3/using/cmdline.html#id3
[`-I`]: https://docs.python.org/3/using/cmdline.html#id2
[`sys.flags`]: https://docs.python.org/3/library/sys.html#sys.flags

Comment (inglesp) [reply django/core/management/commands/shell.py:61]: Thanks for your input.

> Anyway this is still a breaking change... Your patch now only enables macOS support under some operating conditions.

I'm not sure about that, because as far as I can see these operating conditions are only ever theoretical.

If you start Python with `-S`, then you can't load the shell at all:

```
$ python -S manage.py shell
Traceback (most recent call last):
  File "manage.py", line 11, in main
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 22, in <module
    main()
  File "manage.py", line 13, in main
    raise ImportError(
ImportError: Couldn't import Django. Are you sure it's installed and available on your PYTHONPATH environment variable? Did you forget to activate a virtual environment?
```

And if you start it with `-i`, you can start the shell, but you can't import any of your apps:

```
$ python -I manage.py shell
Python 3.8.2 (default, May  2 2020, 15:51:29) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import myapp
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ModuleNotFoundError: No module named 'myapp'
```

I suppose you could argue there is a breaking change if you're somehow relying on an overridden `sys.__interactivehook__()` _not_ being called, but I'm struggling to come up with why you might do that.

> I'm not sure it's a good idea to call `site.enablerlcompleter`: On the one hand, it's not documented. On the other, the name doesn't start with an underscore.

We're not calling `site.enablerlcompleter`, but instead we're calling whatever the value of `sys.__interactivehook__` is, and [that is documented](https://docs.python.org/3/library/sys.html#sys.__interactivehook__).  (I know it begins with an underscore, but I think that's OK.)

> One option might be unconditionally calling `site.main` so that user and site customization script behavior matches vanilla Python.

That's a nice idea, but by the time we get to the code that starts the Django shell, it will already have been called.  In any case, it calls `enablerlcompleter`, which _sets_ `sys.__interactivehook__`.  Something still needs to _call_ `sys.__interactivehook__`, and in normal Python shell, this is done by [code in `main.c`](https://github.com/python/cpython/blob/c9b8e9c421b57acdcaf24fab0c93bc29b3ef7c67/Modules/main.c#L441-L472), which as far as I can tell we can never access from Python.

Comment (wkschwartz) [reply django/core/management/commands/shell.py:61]: > these operating conditions are only ever theoretical.

They're not. I use them all the time together with [`shell`]'s [`--pythonpath`] option. (And I do this on a Mac, hence my interest in making sure we get this PR right.)

[`--pythonpath`]: https://docs.djangoproject.com/en/3.1/ref/django-admin/#cmdoption-pythonpath
[`shell`]: https://docs.djangoproject.com/en/3.1/ref/django-admin/#shell

File: django/core/management/commands/shell.py [hunk 2]
```diff
@@ -79,6 +62,27 @@ def python(self, options):
...
+            traceback.print_exc()
+
+        # Set up tab completion for objects imported by $PYTHONSTARTUP or
+        # ~/.pythonrc.py.
+        try:
+            import readline
+            import rlcompleter
+            readline.set_completer(rlcompleter.Completer(imported_objects).complete)
+        except ImportError:
+            pass
```
Comment (wkschwartz) [inline django/core/management/commands/shell.py:83]: From @inglesp's [comment](https://github.com/django/django/pull/13911/files#r565608154) above:
> We're not calling `site.enablerlcompleter`, but instead we're calling whatever the value of `sys.__interactivehook__` is

My [original mention of the `site` module](https://github.com/django/django/pull/13911/files#r565417290) was ambiguous. What I meant by 
> lean directly on the `site` module

was something like this

```suggestion
        # ~/.pythonrc.py. We intentionally ignore sys.flags.isolated so that
        # readline is available from the command line in otherwise isolated
        # environments for which sys.__interactivehook__ might not already
        # have been set by the site module.
        try:
            import site
        except ImportError:
            pass
        else:
            old_hook = getattr(sys, "__interactivehook__", None)
            try:
                site.enablerlcompleter()
                sys.__interactivehook__()
            finally:
                if old_hook is not None:
                    sys.__interactivehook__ = old_hook
```
Again, I'm not entirely convinced calling `site.enablerlcompleter` itself is advisable. But the point is that at least `site.enablerlcompleter` includes all the libedit compatibility shims.

**Review by wkschwartz — CHANGES_REQUESTED** [review]
I'm concerned that without restoring the libedit compatibility shims, the shell command will lose readline support on macOS in isolated Python setups, which is part of my daily workflow.

**Comment by inglesp** [comment]
> I'm concerned that without restoring the libedit compatibility shims, the shell command will lose readline support on macOS in isolated Python setups, which is part of my daily workflow.

Can you describe your workflow?

**Comment by carltongibson** [comment]
Hey @inglesp @wkschwartz — thanks for the discussion here. Reading it I'm not 100% clear on the resolution. 

> Can you describe your workflow?

@wkschwartz That would be handy. Exactly what are you doing, and how is it broken by this change? If you could specify that then it would really help to know what target we need to hit here. 

Otherwise I fear we're stuck in the long-grass a bit... — not sure how to proceed there. 

I'm going to mark this _Patch needs improvement_ on the ticket so we can get this resolved, but happy to continue discussion. (@inglesp — Also happy for you to make the _It's not broken!_ case if you think we should be moving forward already.)

**Comment by inglesp** [comment]
Thanks @carltongibson.  Coming back to this, I can see it's quite hard to follow!  I'll write up a summary this evening (UK time) and will outline the outstanding questions.

**Comment by carltongibson** [comment]
@inglesp Hero. Thanks!

**Comment by wkschwartz** [comment]
> Can you describe your workflow?

Yes, but it'll be a day or three before I can give this proper attention. In the meantime, please don't merge!

**Comment by inglesp** [comment]
Here's a summary of where I think we're at.  I hope this is clear.

#### Intended changes

This PR attempts to provide out-of-the-box support for `manage.py shell` to preserve command history between sessions.  As an incidental but beneficial side-effect, it also adds support for executing a user-supplied hook when starting a session.

[As of Python 3.4](https://docs.python.org/3/whatsnew/3.4.html), this is the behaviour of the normal Python interactive interpreter.f

#### Unintended changes

However, it breaks tab completion in @wkschwartz's "isolated mode" workflow, although after reading more about isolated mode, I believe that Django's existing behaviour is wrong here, and William is unlucky to be relying on this.

I don't know enough about the rationale behind William's setup to know how common his workflow is, or how many other Django users will be affected.

#### What the PR does

The PR adds a call to `sys.__interactivehook__()` during the setup of a Django shell session, replacing existing code that configures readline directly.

#### About `sys.__interactivehook__`

From the [Python docs about `sys.__interactivehook__`](https://docs.python.org/3/library/sys.html#sys.__interactivehook__):

> When this attribute exists, its value is automatically called (with no arguments) when the interpreter is launched in [interactive mode](https://docs.python.org/3/tutorial/interpreter.html#tut-interactive). This is done after the `PYTHONSTARTUP` file is read, so that you can set this hook there. The `site` module [sets this](https://docs.python.org/3/library/site.html#rlcompleter-config).

In other words, `sys.__interactivehook__` can refer to a function that should be called when the interpreter launches, and it can be set when the `PYTHONSTARTUP` file is run.

If not explicitly set in the `PYTHONSTARTUP` file, and if not disabled by any command line flags (see below), the default value of `sys.__interactivehook__` is [`register_readline`](https://github.com/python/cpython/blob/7247f6f433846c6e37308a550e8e5eb6be379856/Lib/site.py#L443-L489), which (a) sets up tab completion, and (b) reads any history in `~/.python_history`, and (c) registers a function to write to `.python_history` when the interpreter exits.

This default value is set at the end of the call to [`site.enablerlcompleter`](https://github.com/python/cpython/blob/7247f6f433846c6e37308a550e8e5eb6be379856/Lib/site.py#L434-L491), which is called from [`site.main`](https://github.com/python/cpython/blob/7247f6f433846c6e37308a550e8e5eb6be379856/Lib/site.py#L609), as long as `sys.flags.isolated` is not set.

`sys.flags.isolated` is set if Python is run in isolated mode.  From the [Python docs about command line options](https://docs.python.org/3/using/cmdline.html#id2):

> In isolated mode `sys.path` contains neither the script’s directory nor the user’s site-packages directory. All `PYTHON*` environment variables are ignored, too. Further restrictions may be imposed to prevent the user from injecting malicious code.

One of these futher restrictions is that `readline` is not imported, "as it could conceivably be arbitrary code".  See [this BPO issue](https://bugs.python.org/issue28192) and [this commit](https://github.com/python/cpython/commit/313523ce2dd67db67b7dcbcc5f6d6ca95922b34b).

#### About Django and isolated mode

The Django shell can be started by running in isolated mode, with `python -I ./manage.py shell --pythonpath .`.

I'm unclear on the benefits to doing this, and I hope that @wkschwartz can shed some light here.

At the moment, when the Django shell is started in isolated mode, tab completion is enabled.  This is because the existing code in [django/core/management/commands/shell.py](https://github.com/django/django/blob/main/django/core/management/commands/shell.py#L48-L63) doesn't consider isolated mode.  I actually think this is a mistake, since it circumvents some of the protection that isolated mode should provide.  This is because the readline support was added to Django [in 2006](https://github.com/django/django/commit/427a54b4a369b160f9d0ce76fdd72bb3cd89f142), while isolated mode was introduced in Python 3.4, [from 2014](https://docs.python.org/3/whatsnew/3.4.html).

Given the restrictions of isolated mode, I'm not aware of any way that tab completion can be enabled before the interpreter starts, apart from the current implementation.

#### Alternatives

There are a couple of changes that would add support for `.python_history` while maintaining `readline` support in isolated mode.

1. We could call `site.enablerlcompleter()` explicitly from the setup in `shell.py`.  However, doing this means that a user cannot provide their own version of `sys.__interactivehook__`, and means that Django is depending on something undocumented.
2. We could copy the existing code in the `register_readline` function into Django's `shell.py`.  This is more or less what we currently do, except that the copied code is from an old version of `register_readline` and so is missing the changes that were added to write to `.python_history` at the end of a session.  The risk here is that the copied code gets out of date again.

However in both cases, we would be circumventing some of the protection that isolated mode is supposed to provide.

We could also document that users can provide a `PYTHONSTARTUP` file that passes a path to `readline.read_history_file` and sets an `atexit` hook to write it at the end of a session.  This is what the [Python docs recommended](https://docs.python.org/3.3/tutorial/interactive.html) before 3.4.  It feels like a backwards step to recommend it to Django users today.

#### Next steps

* We need to make a decision about whether it is acceptable to break tab completion for users of isolated mode.
* If it's not acceptable to break this, then we need to make a decision about whether we want to support preserving history between sessions.
* If we want to support that, then we need to decide which of the alternative implementations is most suitable, or to come up with another one.

@wkschwartz, it would be helpful if you could add some more information about your workflow.

**Comment by carltongibson** [comment]
Thanks @inglesp — let me have a read/think. 👍

**Comment by wkschwartz** [comment]
Thank you,  @inglesp, for your thorough [write up](https://github.com/django/django/pull/13911#issuecomment-862764723). Upon further consideration, let me refine my position: **The PR's basic approach is fine; it's just a sufficiently breaking change that it should occur only in a major release.** Since [`main` is currently for Django 4.0](https://github.com/django/django/blob/225d96533a8e05debd402a2bfe566487cc27d95f/django/__init__.py#L3), **now is a fine time to merge it** (after whatever further review/editing is needed). 

> Django's existing behaviour is wrong here, and William is unlucky to be relying on this.

I take the point that the part of the purpose of isolated mode is to turn off automatically setting up `readline`. Having `python -m django shell` work as similarly as possible to just plain `python` makes sense to me (and I particularly value that `shell` has a `-c`!). 

My use case is admittedly a bit niche, but it probably makes more sense for my code to make an exception to its `-I -S` policy for `shell` than for `shell` to work so differently from the normal Python REPL. My Django project is [compiled](https://pyoxidizer.readthedocs.io/) into a so-called "frozen" Python. `sys.path` is hard coded with relative paths to a `site-packages` directory containing Django. The frozen Python is then [configured](https://docs.python.org/3/c-api/init_config.html#pyconfig) to act as though it's [starting](https://docs.python.org/3.9/using/cmdline.html) with `-I -E -S -s -X utf8`. We then have scripts that call the executable with Django's `--pythonpath` if we need to run tests or inject any other dependencies. If we're running `shell` it's for debugging or testing, and strict isolation isn't as important as it is in production. Django's current behavior of unconditionally setting up readline was nice (though, now that I'm thinking of it, was always pretty flaky in the frozen Python form for reasons that I think have nothing to do with Django). After this PR, I'll need to add a shim to my app that intercepts `shell`, calls `site.enablerlcompleter()`, and allows `shell` to  continue by calling `sys.__interactivehook__()`.

I don't know how many users have this set up, but [it's possible that it's not zero](https://code.djangoproject.com/ticket/30950). If the change goes into Django 4.0 with a line in the release notes, then those of us who relied on the existing behavior can modernize.
There are dozens of us!

**Comment by carltongibson** [comment]
Thank you both for your input! I'll take a look at this again next week to read through all your comments, and make sure I'm up on it. Good work!

**Comment by carltongibson** [comment]
Hey @inglesp and @wkschwartz — thanks for your working together to resolve this. Good research. 

> I take the point that the part of the purpose of isolated mode is to turn off automatically setting up readline. ... it probably makes more [...adjust current usage...] than for `shell` to work so differently from the normal Python REPL.

I think that's the essence. Good 👍 — I'll look to give this the final review now.

File: django/core/management/commands/shell.py (hunk 2 above)
Comment (carltongibson) [inline django/core/management/commands/shell.py:83]: Hey @inglesp — can you crib for me why we need this block? 🤔 Thanks.

Comment (carltongibson) [reply django/core/management/commands/shell.py:83]: It's OK, I see — `imported_objects` is used as the interpreter environment, and we `exec` `.pythonrc` into it. — I think I was looking at the wrong diff. 😜

File: docs/releases/4.0.txt [hunk 3]
```diff
@@ -236,6 +236,11 @@ Management Commands
 
 * On PostgreSQL, :djadmin:`dbshell` now supports specifying a password file.
 
+* Internally the :djadmin:`shell` command now respects
+  :py:data:`sys.__interactivehook__` at startup. Primary this allows loading
+  shell history between interactive sessions. As a consequence, ``readline`` is
+  no longer loaded if running in *isolated* mode.
```
Comment (wkschwartz) [inline docs/releases/4.0.txt:242]: I find cross reference links extremely helpful. Others find them unnecessary. Up to y'all.

```suggestion
  :ref:`shell history <python:tut-keybindings>` between interactive sessions. As a consequence, :py:mod:`readline` is
  no longer loaded if running in :ref:`isolated <python:init-isolated-conf>` (:option:`python:-I`) mode.
```

**Review by wkschwartz — COMMENTED** [review]
Just some minor stuff.

File: docs/releases/4.0.txt [hunk 4]
```diff
@@ -236,6 +236,11 @@ Management Commands
 
 * On PostgreSQL, :djadmin:`dbshell` now supports specifying a password file.
 
+* Internally the :djadmin:`shell` command now respects
+  :py:data:`sys.__interactivehook__` at startup. Primary this allows loading
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply docs/releases/4.0.txt]: Yep, good spot. 👍

File: docs/releases/4.0.txt (hunk 3 above)
Comment (carltongibson) [reply docs/releases/4.0.txt:242]: This seems a bit overkill to me here. 

Whilst linking linking to `sys.__interactivehook__` is something that folks might learn something from here, I don't take it that we're teaching about these other items. Folks know what shell history and `readline` are, and we're not really advocating isolated mode. (Rather making sure you know we changed it.) 

But 🤷

**Comment by inglesp** [comment]
Hi both -- I've been on holiday this week, and I'm delighted to see this has been resolved while I've been away.  Thanks!

**Comment by carltongibson** [comment]
Thanks for your effort @inglesp it's a good addition!

**Comment by timgraham** [comment]
I haven't read through all the history here, but is it expected that some users will now see something like this?
```
$ python manage.py shell
Failed calling sys.__interactivehook__
Traceback (most recent call last):
  File "/home/tim/code/django/django/core/management/commands/shell.py", line 69, in python
    sys.__interactivehook__()
AttributeError: module 'sys' has no attribute '__interactivehook__'
Python 3.9.5 (default, May 21 2021, 18:04:17) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>>
```
As a user, I'm not sure what action to take (if any) to suppress the "Failed calling sys.\_\_interactivehook\_\_" message.

**Comment by inglesp** [comment]
Hi Tim

> is it expected that some users will now see something like this?

Er, no it's not!  I've tried to reproduce with the same version of Python (3.9.5) on linux, and it works as expected.

```
$ python --version
Python 3.9.5
$ pip freeze
asgiref==3.4.1
-e git+git@github.com:django/django.git@b8824e8d17434957650517370f739e340867c381#egg=Django
pytz==2021.1
sqlparse==0.4.1
```

What happens if you start a normal Python shell in the same virtualenv?  And from that normal Python shell, what does `sys.__interactivehook__` give you?

**Comment by timgraham** [comment]
A normal Python shell doesn't give an error.
`sys.__interactivehook__` gives `AttributeError`.

For the problematic virtualenvs, the `site.py` in the virtualenv doesn't have the `enablerlcompleter()` function that sets  `sys.__interactivehook__`.

**Comment by inglesp** [comment]
Thanks, that's surprising.

How do you create your virtualenvs?  And can you upload a copy of a `site.py` that doesn't have the `enablerlcompleter()` function?

(Also: is this PR the best place for this discussion?)

**Comment by timgraham** [comment]
I use mkvirtualenv to create the env. I think the issue was an older version of virtualenv (I had 15.2.0 installed) which I think either bundles its own `site.py` or was using one from Python 2 with some modifications. I upgraded to the latest virtualenv and `site.py` is no longer created in the virtualenv so the one from the base Python is used.

Anyway, it might be worth adding a guard like `if hasattr(sys, '__interactivehook__'):` before making the call.

**Comment by inglesp** [comment]
Thanks, that all makes sense.  Reading the Python source more carefully, I see that it handles the case where `sys.__interactivehook__` is missing.  We should do the same too.

Should I create a new PR, or does this need to go via a ticket?

**Comment by timgraham** [comment]
It should be okay to reference the original ticket since the issue is fairly straightforward and this hasn't been released yet.

**Comment by inglesp** [comment]
Thanks, see #14837.
